src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,166 @@
+  const auto segment_size = segment.size();
+
+  // Note: If _seach_value is larger than all values in the dictionary segment, then the ids returned are
+  //       INVALID_VALUE_ID - the biggest possible value of any ValueID, so the math checks out.
+  const auto comparison_value =
+      InBetweenValueID{segment.lower_bound(_search_value), segment.upper_bound(_search_value)};
+
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < segment_size; ++chunk_offset) {
+    if (predicate(attribute_vector->get(chunk_offset), comparison_value)) {
+      pos_list.push_back({chunk_id, chunk_offset});
+    }
+  }
+}
+
+void TableScan::_scan_reference_segment(const ReferenceSegment& segment, PosList& pos_list) {
+  const auto predicate = predicate_for_scantype<AllTypeVariant>(_scan_type);
+  for (const auto row_id : *segment.pos_list()) {
+    const auto value = segment.get_by_row_id(row_id);
+    if (!variant_is_null(value) && predicate(value, _search_value)) {
+      pos_list.emplace_back(row_id);
+    }
+  }
+}
+
+ColumnID TableScan::column_id() const {
+  return _column_id;
+}
+
+ScanType TableScan::scan_type() const {
+  return _scan_type;
+}
+
+const AllTypeVariant& TableScan::search_value() const {
+  return _search_value;
+}
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+  const auto input_table = _left_input_table();
+  const auto chunk_count = input_table->chunk_count();
+
+  // When scanning a ReferenceSegment, the newly created ReferenceSegment should not point into the former, but instead
+  // refer to the underlying Table directly.
+  auto referenced_table = input_table;
+  auto pos_list = std::make_shared<PosList>();
+  auto reference_segment_count = 0;
+
+  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+    const auto chunk = input_table->get_chunk(chunk_id);
+    const auto segment = chunk->get_segment(_column_id);
+
+    resolve_data_type(input_table->column_type(_column_id), [&](auto type) {
+      using Type = typename decltype(type)::type;
+      const auto value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment);
+      const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment);
+      const auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+
+      DebugAssert(value_segment || dictionary_segment || reference_segment,
+                  "Segment has to be ValueSegment, DictionarySegment or ReferenceSegment.");
+
+      if (value_segment) {
+        _scan_value_segment(chunk_id, *value_segment, *pos_list);
+      } else if (dictionary_segment) {
+        _scan_dictionary_segment(chunk_id, *dictionary_segment, *pos_list);
+      } else if (reference_segment) {
+        _scan_reference_segment(*reference_segment, *pos_list);
+        referenced_table = reference_segment->referenced_table();
+        ++reference_segment_count;
+      }
```

Comment:
How so? How do you think it can be improved?

Reply:
Sorry, this comment seems to have slipped through. We meant to remove it as we could not think of better ways to structure it. The initial comment was made by me due to the placement of the assert between casts and accesses. It could also be moved to another else-branch with a `Fail()` but we decided to omit the comment.